pkg/rtc/participant.go

```diff
@@ -447,11 +421,10 @@ func (p *ParticipantImpl) Close() error {
 	onClose := p.onClose
 	p.lock.RUnlock()
 	if onClose != nil {
-		onClose(p)
+		onClose(p, disallowedSubscriptions)
 	}
 	p.publisher.Close()
 	p.subscriber.Close()
-	close(p.rtcpCh)
```

Comment:
I believe this needs to be closed last, after the downtracks are closed. this is to prevent downtracks from writing additional rtcp messages into the channel after channel closure.

Reply:
This is used only for the publisher side. So, I moved this variable into `UptrackManager` and closing it there. The `UptrackManager.Close()` is called ☝️ . Do you think that needs to be moved here?

Reply:
where it gets called matters isn't the issue.. but the *order* that it's called.

if rtcpCh is closed **before** downtracks, then downtracks could still write rtcp messages into the channel. In Go, writing into closed channel causes a panic.

hence.. we have to close all of the downtracks **before** the uptrack

Reply:
Sorry for being dense here @davidzhao . I am not able to understand the connection to down tracks here.

`rtcpCh` is created in `UptrackManager` and it sets up a worker routine process RTCP packets from that channel and calls back to write publisher peer connection.

`rtcpCh` is given to `MediaTrack` and `MediaTrack` uses that in two places
- https://github.com/livekit/livekit-server/blob/eae6eff6a3e41cf6c49bc8bcfd959b1c79f2fd69/pkg/rtc/mediatrack.go#L417 - which is used in `sfu.Receiver` here https://github.com/livekit/livekit-server/blob/eae6eff6a3e41cf6c49bc8bcfd959b1c79f2fd69/pkg/sfu/receiver.go#L390. There could be PLIs from the downtracks, but those are down tracks of the subscribing participants and not of this participant.
- https://github.com/livekit/livekit-server/blob/eae6eff6a3e41cf6c49bc8bcfd959b1c79f2fd69/pkg/rtc/mediatrack.go#L656 - this is in the feedback handler when `sfu.Buffer` sends RTCP feedback.

And that feedback from `sfu.Buffer` comes from only two places
- https://github.com/livekit/livekit-server/blob/eae6eff6a3e41cf6c49bc8bcfd959b1c79f2fd69/pkg/sfu/buffer/buffer.go#L398 - when sending NACK for upstream losses
- https://github.com/livekit/livekit-server/blob/eae6eff6a3e41cf6c49bc8bcfd959b1c79f2fd69/pkg/sfu/buffer/buffer.go#L422 which gets packets from https://github.com/livekit/livekit-server/blob/eae6eff6a3e41cf6c49bc8bcfd959b1c79f2fd69/pkg/sfu/buffer/buffer.go#L525 and those are purely stuff pertaining to the uptrack (RTCP receiver reports and REMB or TWCC)

In `UptrackManager.Close`, `RemoveAllSubscibers` is called before closing the channel.

But, I do see an issue of `RemoveAllSubscribers` spawning go routines to close the down tracks, but we are synchronously closing the channel. But, that pattern has not changed (before the change up tracks `RemoveAllSubscribers` with some spacing from `rtcpCh` close, but I am not sure if that is enough to ensure that the channel is not closed before all down tracks are closed from all the spawned go routines to close the down tracks). Looks to me the right way to do it would be to close the channel in up track close callback by checking that there are no published/pending tracks left.

I am definitely lacking some understanding here.